File: Test Case Generation/gemini/cpp/test_task_17.cpp
#include <gtest/gtest.h>
#include <vector>
#include <string>

// Protótipo da função a testar
std::vector<int> parse_music(std::string music_string);

// Teste para garantir que uma string vazia retorna um vetor vazio
TEST(ParseMusicTest, HandlesEmptyInput) {
    std::string input = "";
    std::vector<int> expected = {};
    EXPECT_EQ(parse_music(input), expected);
}

// Teste do exemplo fornecido na documentação
TEST(ParseMusicTest, HandlesStandardInput) {
    std::string input = "o o| .| o| o| .| .| .| .| o o";
    std::vector<int> expected = {4, 2, 1, 2, 2, 1, 1, 1, 1, 4, 4};
    EXPECT_EQ(parse_music(input), expected);
}

// Teste individual para cada tipo de nota
TEST(ParseMusicTest, HandlesIndividualNotes) {
    EXPECT_EQ(parse_music("o"), std::vector<int>{4});
    EXPECT_EQ(parse_music("o|"), std::vector<int>{2});
    EXPECT_EQ(parse_music(".|"), std::vector<int>{1});
}

// Teste com múltiplas instâncias da mesma nota
TEST(ParseMusicTest, HandlesRepeatedNotes) {
    std::string input = "o o o";
    std::vector<int> expected = {4, 4, 4};
    EXPECT_EQ(parse_music(input), expected);
}

// Teste com espaços extras (conforme a lógica da função, espaços duplos resultam em 'current' vazio)
TEST(ParseMusicTest, HandlesExtraSpaces) {
    std::string input = "o  o|"; 
    std::vector<int> expected = {4, 2};
    EXPECT_EQ(parse_music(input), expected);
}

// Teste com caracteres inválidos ou formatos não reconhecidos (deve ignorar e não adicionar ao vetor)
TEST(ParseMusicTest, HandlesInvalidNotes) {
    std::string input = "o invalid .| x o|";
    std::vector<int> expected = {4, 1, 2};
    EXPECT_EQ(parse_music(input), expected);
}

// Teste de cobertura de fluxo: string com apenas espaços
TEST(ParseMusicTest, HandlesOnlySpaces) {
    std::string input = "   ";
    std::vector<int> expected = {};
    EXPECT_EQ(parse_music(input), expected);
}

// Teste para garantir que notas sem espaço no final são processadas (Branch Coverage do loop final)
TEST(ParseMusicTest, HandlesNoTrailingSpace) {
    std::string input = "o| .| o";
    std::vector<int> expected = {2, 1, 4};
    EXPECT_EQ(parse_music(input), expected);
}

File: Test Case Generation/gemini/python/test_task_17.py
import pytest
from typing import List

# A função a ser testada (incluída para garantir que os testes correm autonomamente)
def parse_music(music_string: str) -> List[int]:
    note_map = {'o': 4, 'o|': 2, '.|': 1}
    return [note_map[x] for x in music_string.split(' ') if x]

def test_parse_music_basic_example():
    """Testa o exemplo fornecido no docstring."""
    input_str = 'o o| .| o| o| .| .| .| .| o o'
    expected = [4, 2, 1, 2, 2, 1, 1, 1, 1, 4, 4]
    assert parse_music(input_str) == expected

def test_parse_music_empty_string():
    """Testa uma string vazia, que deve retornar uma lista vazia."""
    assert parse_music('') == []

def test_parse_music_single_notes():
    """Testa cada nota individualmente."""
    assert parse_music('o') == [4]
    assert parse_music('o|') == [2]
    assert parse_music('.|') == [1]

def test_parse_music_multiple_spaces():
    """Testa o comportamento com espaços extras (deve ignorar devido ao 'if x')."""
    input_str = 'o  o|   .|'
    expected = [4, 2, 1]
    assert parse_music(input_str) == expected

def test_parse_music_leading_trailing_spaces():
    """Testa espaços no início e no fim da string."""
    assert parse_music(' o ') == [4]
    assert parse_music('  o| .|  ') == [2, 1]

def test_parse_music_all_same_notes():
    """Testa uma sequência da mesma nota."""
    assert parse_music('o o o') == [4, 4, 4]
    assert parse_music('.| .| .| .|') == [1, 1, 1, 1]

def test_parse_music_invalid_note():
    """Testa o comportamento ao encontrar uma nota inválida (deve lançar KeyError)."""
    with pytest.raises(KeyError):
        parse_music('o x| .|')

def test_parse_music_invalid_format_no_spaces():
    """Testa notas coladas sem espaços (deve lançar KeyError pois 'oo|' não existe no mapa)."""
    with pytest.raises(KeyError):
        parse_music('oo|')

def test_parse_music_newline_tabs():
    """Testa se outros caracteres de whitespace (não espaços) causam erro ou comportamento inesperado."""
    # Como o split é explicitamente por ' ', um \n resultaria em uma chave inválida 'o\no'
    input_with_newline = 'o\no'
    with pytest.raises(KeyError):
        parse_music(input_with_newline)